src/mpid/ch4/netmod/ofi/ofi_events.c

```diff
@@ -321,9 +321,9 @@ int MPIDI_OFI_get_huge_event(struct fi_cq_tagged_entry *wc, MPIR_Request * req)
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_OFI_GET_HUGE_EVENT);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_OFI_GET_HUGE_EVENT);
 
-    if (recv_elem->localreq && recv_elem->cur_offset != 0) {    /* If this is true, then the message has a posted
-                                                                 * receive already and we'll be able to find the
-                                                                 * struct describing the transfer. */
+    /* If this is true, then the message has a posted receive already and we'll be able to find the
+     * struct describing the transfer. */
+    if (recv_elem->localreq && recv_elem->cur_offset != 0 && recv_elem->remote_info.msgsize != 0) {
```

Comment:
Do we understand when `localreq` will be set or unset, when `cur_offset` will be `0` or non-zero, and `when `remote_info.msgsize` will be `0` or non-zero?

Right now reading this code, we are looking at (2*2*2 = 8) possibilities and it requires tracing a lot of code to understand. If we can narrow down the condition and either refactor the condition or document the conditions (causes of the conditions), that will help the maintainence.

Reply:
> Do we understand when `localreq` will be set or unset, 

When the message is posted. 

> when `cur_offset` will be `0` or non-zero, 

When the first part of the message has arrived. 

> and `when `remote_info.msgsize` will be `0` or non-zero?

Confirming that we didn't underflow. We might be able to drop this part now that we're adding the previous condition.

Reply:
This isn't about handling lots of cases. We only want to handle one and drop all of the rest.

Reply:
Thanks. That is helpful. I guess my real question is, what are the cases that we drop?

Reply:
All the cases where a message hasn't been matched yet (e.g, unexpected only or posted but not received).

Reply:
Alright, I spent my time understanding the logic, cross-referencing all the relevant functions. My understanding results in #5364. 

When reading `MPIDI_OFI_get_huge_event`, the natural question (as I asked) is when this function is being called (its roles/purposes), and if this function only processes some cases, what happens to the other cases. Both questions can't be understood without reading all other code and put all the code together.

Now, if we move the condition to the caller site, the context makes it really simple. When recv is received, we check whether control message has been recved. If yes, start `MPIDI_OFI_get_huge_event`; otherwise, leave for the control message handler to do that.

When the control message handler is called, we check whether receive has been posted. If yes, start `MPIDI_OFI_get_huge_event`; otherwise, leave for the recv to do that.

Both logic is straightforward and can be understood by just understand locally. In particular, I don't need read a single line in `MPIDI_OFI_get_huge_event` to understand it.

Reply:
It was clear enough to me before, but if you prefer to refactor things, I won't stop you...